src/QsCompiler/SyntaxProcessor/CapabilityInference.fs

```diff
@@ -193,48 +196,85 @@ let private scopePatterns scope = scope.Statements |> Seq.collect statementPatte
+    let visited = Set.add spec.Parent visited
+    let inOperation = callables.TryGetValue spec.Parent |> tryToOption |> Option.exists isOperation
+    let dependencies =
+        graph.GetDirectDependencies spec
+        |> Seq.map (fun group -> group.Key.CallableName)
+        |> Set.ofSeq
+        |> fun names -> Set.difference names visited
+    dependencies
+    |> Seq.choose (callables.TryGetValue >> tryToOption)
+    |> Seq.map (callableDependentCapability callables graph visited)
+    |> Seq.append (specSourceCapability inOperation spec |> Seq.singleton)
+    |> maxCapability
+
+/// Returns the required capability of the callable based on its capability attribute if one is present. If no attribute
+/// is present and the callable is not defined in a reference, returns the capability based on its source code and
+/// callable dependencies. Otherwise, returns the base capability.
+and private callableDependentCapability callables graph visited (callable : QsCallable) =
+    callable.Attributes
+    |> QsNullable<_>.Choose BuiltIn.GetCapability
+    |> tryMaxCapability
+    |> Option.defaultWith (fun () ->
+        if callable.SourceFile.Value.EndsWith ".qs"
+        then
+            callable.Specializations
+            |> Seq.map (specDependentCapability callables graph visited)
+            |> maxCapability
+        else baseCapability)
+
+/// Adds the given capability to the callable as an attribute.
+let private addAttribute callable capability =
     let arg = capability.ToString () |> AttributeUtils.StringArgument
     let attribute = AttributeUtils.BuildAttribute (BuiltIn.Capability.FullName, arg)
-    { callable with Attributes = callable.Attributes.Add attribute }
+    { callable with QsCallable.Attributes = callable.Attributes.Add attribute }
 
 /// Infers the capability of all callables in the compilation, adding the built-in Capability attribute to each
 /// callable.
-///
-/// TODO:
-/// This infers the capability of each callable individually, without looking at the capabilities of anything it
-/// calls or references. The inferred capability is only a lower bound.
 let InferCapabilities compilation =
+    let callables = GlobalCallableResolutions compilation.Namespaces
+    let graph = BuildCallGraph.Apply compilation
     let transformation = SyntaxTreeTransformation ()
     transformation.Namespaces <- {
         new NamespaceTransformation (transformation) with
             override this.OnCallableDeclaration callable =
-                if callable.Attributes |> QsNullable<_>.Choose BuiltIn.GetCapability |> Seq.isEmpty
-                then addAttribute callable
+                let isMissingCapability =
+                    callable.Attributes |> QsNullable<_>.Choose BuiltIn.GetCapability |> Seq.isEmpty
+                if isMissingCapability && callable.SourceFile.Value.EndsWith ".qs"
+                then callableDependentCapability callables graph Set.empty callable |> addAttribute callable
```

Comment:
If you could keep track of the visited callables as well as what capability is for visited callables here, you could check for the "new" version of callables, or at least to see the capability of visited callables. This would prevent the transformation from having to look down the same lineages of dependencies multiple times.

Reply:
Perhaps instead of a set of QsQualified names, the data type could be a map of the visited names to the capability found for each.

Reply:
It needs to add a name to `visited` at the start of the function, before the capability has been computed, to prevent infinite loops with cycles, so I don't think `visited` can be turned into a map. However, I did add a separate map to store computed capabilities. As a result of that I also needed to handle cycles separately. The new algorithm should avoid recomputing capabilities.